src/main/java/org/elasticsearch/index/translog/fs/ChannelReference.java

```diff
@@ -56,8 +55,54 @@ public FileChannel channel() {
         return this.channel;
     }
 
+    public TranslogStream stream() {
+        return this.stream;
+    }
+
+    /**
+     * called to add this owner to the list of reference holders (used for leakage detection).
+     * also asserts that there is no double "attachment"
+     */
+    boolean assertAttach(FsChannelReader owner) {
+        Map<FsChannelReader, RuntimeException> ownerMap = openedFiles.get(file().toString());
+        Throwable previous = ownerMap.put(owner, new RuntimeException(file.toString() + " attached", null));
+        assert previous == null : "double attachment by the same owner";
+        return true;
+    }
+
+    /** removes an owner to the least of list holders (used for leakage detection).
+     * also asserts that this owner did attach before.
+     */
+    boolean assertDetach(FsChannelReader owner) {
+        Map<FsChannelReader, RuntimeException> ownerMap = openedFiles.get(file().toString());
+        Throwable previous = ownerMap.remove(owner);
+        assert previous != null : "reader detaches, but was never attached";
+        return true;
+    }
+
+    @Override
+    public String toString() {
+        return "channel: file [" + file + "], ref count [" + refCount() + "]";
+    }
+
     @Override
     protected void closeInternal() {
         IOUtils.closeWhileHandlingException(channel);
+        assert openedFiles.remove(file().toString()) != null;
+    }
+
+    // per file, which objects refer to it and a throwable of the allocation code
+    static final Map<String, Map<FsChannelReader, RuntimeException>> openedFiles;
+
+    static {
+        boolean assertsEnabled = false;
+        assert (assertsEnabled = true);
+        if (assertsEnabled) {
+            openedFiles = ConcurrentCollections.newConcurrentMap();
```

Comment:
I spoke to @rmuir and this is enabled 99% of the time in master (file leak detection) we can just remove this layer.